tests/functional/verification/test_reflected_magic_func.py

```diff
@@ -0,0 +1,278 @@
+    Requires(False)
+    assert False
+
+class I1(H1):
+  def __radd__(self, other: 'H1') -> 'H1':
+    return I1()
+
+def incompatible_add_test() -> None:  
+  r51 = H1() + I1()
+
+
+############################################
+##  Subtyped operands                     ##
+############################################
+
+# If operand inherits magic functions from supertype
+# without overriding them, treat operand as usual
+class J():
+  def __add__(self, other: int) -> 'J':
+    Requires(False)
+    assert False
+  
+  def __radd__(self, other: int) -> 'J':
+    return J()
+  
+class K(J):
+  pass
+
+def basic_subclass_test() -> None:
+  r6 = 2 + K()
+
+class L():
+  def __add__(self, other: int) -> 'L':
+    return L()
+  
+  def __radd__(self, other: int) -> 'L':
+    Requires(False)
+    assert False
+  
+class M(L):
+  pass
+def basic_subclass_test_2() -> None:
+  r7 = M() + 2
+
+class P():
+  def __add__(self, other: 'P') -> 'P':
+    return P()
+  
+  def __radd__(self, other: 'P') -> 'P':
+    Requires(False)
+    assert False
+  
+class Q(P):
+  pass
+
+def basic_subclass_test_3() -> None:
+  r9 = P() + Q()
+
+class R():
+  pass
+  
+class S(R):
+  def __radd__(self, other: 'R') -> 'R':
+    return R()
+
+def basic_subclass_test_4() -> None:
+  r10 = R() + S()
+
+
+# If right operand is subtype of left operand, and
+# right operand provides its own implementation of
+# __radd__, prefer this reflected function before
+# __add__ from left operand
+class N():
+  def __add__(self, other: 'N') -> 'N':
+    Requires(False)
+    assert False
+  
+  def __radd__(self, other: 'N') -> 'N':
+    Requires(False)
+    assert False
+  
+class O(N):
+  def __radd__(self, other: 'N') -> 'N':
+    return O()
+  
+def overridden_radd_test() -> None:
+  r8 = N() + O()
+
+class T():
+  pass
+
+class U(T):
+  def __radd__(self: 'U', other: 'U') -> 'U':
+    Requires(False)
+    assert False
+  
+class V(U):
+  def __radd__(self: 'V', other: 'T') -> 'V':
+    return self
+
+def overridden_radd_test_2() -> None:
+  r11 = U() + V()
+
+class W():
+  pass
+
+class X(W):
+  def __add__(self: 'X', other: 'X') -> 'X':
+    Requires(False)
+    assert False
+
+  def __radd__(self: 'X', other: 'X') -> 'X':
+    Requires(False)
+    assert False
+  
+class Y(X):
+  def __radd__(self: 'Y', other: 'W') -> 'Y':
+    return self
+
+def overridden_radd_test_3() -> None:
+  r12 = X() + Y()
+
+
+class IntContainer():
+  def __init__(self, i: int) -> None:
+    self.i = i
+    Ensures(Acc(self.i) and self.i == i)
+
+  def __add__(self, other: 'IntContainer') -> 'IntContainer':
+    Requires(Acc(self.i, 1/2) and Acc(other.i, 1/2))
+    Ensures(Acc(self.i, 1/2) and Acc(other.i, 1/2))
+    Ensures(Acc(Result().i) and Result().i == self.i + other.i)
+    return IntContainer(self.i + other.i)
+  
+
+def failure() -> None:
+  a = IntContainer(42)
+  b = IntContainer(15)
+  c = a + b
+
+  assert c.i == 57
+  #:: ExpectedOutput(assert.failed:assertion.false)
```

Comment:
I've added a few cases that are expected to produce verification failures. They're not much more that some basic smoke tests. Did you have anything more specific in mind?

Reply:
No, this is exactly what I meant, thanks. Just something to make sure e.g. the correct preconditions are actually being checked and we'd catch it if there was an accidental ``assume false`` somewhere in the encoding.